### PR TITLE
修复QA-Engineer在RunCode(执行用例)结束后，要求Engineer WriteCode(修改代码）, Engineer未修改，业务流直接结束的问题

### PR DESCRIPTION
**Pull request by bobooscar**
### **User description**
**Features**
修复QA-Engineer在RunCode(执行用例)结束后，要求Engineer WriteCode(修改代码）, Engineer未修改，业务流直接结束的问题
    
**Influence**
导致过早结束业务流程，Engineer代码有问题而直接结束流程；修复后会增加代码修改流程，增加成功率，但同时会降低完成率。请慎重评估是否合入。

**Result**
修改后，在QA测试发现原码有问题时，Engineer继续修改代码，进行重试。

___

### **PR Type**
Bug fix, Enhancement

___

### **Description**
- Added and enhanced debug and warning logs across multiple roles to improve traceability and debugging.
- Increased the test rounds allowed for QA engineers to ensure thorough testing and bug fixing.
- Included `RunCode` in the actions that can trigger code writing, enhancing the engineer's role responsiveness.

___

### **Changes walkthrough** 📝
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement
</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>base_llm.py</strong><dd><code>Add Debug Logging for LLM Response</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
metagpt/provider/base_llm.py

- Added debug logging for the response from the language model.

</details>
    

  </td>
  <td><a href="https://github.com/geekan/MetaGPT/pull/1267/files#diff-7b80d6287bf4cf75c4061b77e9af5c5e017ba6d774301fa43823e6997c6b423f">+1/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>engineer.py</strong><dd><code>Enhance Logging and Action Handling in Engineer Role</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
metagpt/roles/engineer.py

<li>Added import for <code>RunCode</code>.<br> <li> Added warning logs for cases where no action is determined or nothing <br>gets done.<br> <li> Included <code>RunCode</code> in the set of actions that trigger writing code.<br>

</details>
    

  </td>
  <td><a href="https://github.com/geekan/MetaGPT/pull/1267/files#diff-670e1154f69ca119cd089e5aa9dd579e771211cddcffec1ef1fb66637e385b2a">+4/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>qa_engineer.py</strong><dd><code>Increase QA Engineer Test Rounds Limit</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
metagpt/roles/qa_engineer.py

- Increased the allowed test rounds from 5 to 20.

</details>
    

  </td>
  <td><a href="https://github.com/geekan/MetaGPT/pull/1267/files#diff-aec52f9a5a8049637d41195b2837722f211e2aff9e96766c6ec7aeafb70fd80b">+1/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>role.py</strong><dd><code>Improve Logging for Role Actions and Decisions</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
metagpt/roles/role.py

<li>Added debug logging for actions handling and process flow.<br> <li> Added warning log when no actionable todo is present.<br>

</details>
    

  </td>
  <td><a href="https://github.com/geekan/MetaGPT/pull/1267/files#diff-57eb0c9c5adc353ba3e4db36bb4443d7b86c62b79bd4a084fa105674a253be0d">+3/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    
</table></td></tr></tr></tbody></table>

___

> 💡 **PR-Agent usage**:
>Comment `/help` on the PR to get a list of all available PR-Agent tools and their descriptions

### CODE REVIEW

**Comment by iorisa** [comment]
`Engineer`并未订阅`RunCode`消息，因此即使消息发给了`Engineer`，在`_observe`阶段也会被忽略掉。
建议参考Product Manager给Engineer转bug单的流程，将需要Engineer重写的代码作为bug单转给Engineer。参见`metagpt/actions/write_prd.py`的`_handle_bugfix`：
```python
    async def _handle_bugfix(self, req: Document) -> Message:
        # ... bugfix logic ...
        await self.repo.docs.save(filename=BUGFIX_FILENAME, content=req.content)
        await self.repo.docs.save(filename=REQUIREMENT_FILENAME, content="")
        return AIMessage(
            content=f"A new issue is received: {BUGFIX_FILENAME}",
            cause_by=FixBug,
            send_to="Alex",  # the name of Engineer
        )
```
`metagpt/roles/engineer.py`中的`_think`:
```python
        if self.config.inc and msg.cause_by in write_plan_and_change_filters:
            logger.debug(f"TODO WriteCodePlanAndChange:{msg.model_dump_json()}")
            await self._new_code_plan_and_change_action(cause_by=msg.cause_by)
            return self.rc.todo
```

File: metagpt/provider/base_llm.py [hunk 1]
```diff
@@ -148,6 +148,7 @@ async def aask(
             message.extend(msg)
         logger.debug(message)
         rsp = await self.acompletion_text(message, stream=stream, timeout=self.get_timeout(timeout))
+        logger.debug(f"llm {rsp=}")
```
Comment (iorisa) [inline metagpt/provider/base_llm.py:151]: 这些调试用的代码在提交前可以删掉了

**Comment by bobooscar** [comment]
> `Engineer`并未订阅`RunCode`消息，因此即使消息发给了`Engineer`，在`_observe`阶段也会被忽略掉。 建议参考Product Manager给Engineer转bug单的流程，将需要Engineer重写的代码作为bug单转给Engineer。参见`metagpt/actions/write_prd.py`的`_handle_bugfix`：
> 
> ```python
>     async def _handle_bugfix(self, req: Document) -> Message:
>         # ... bugfix logic ...
>         await self.repo.docs.save(filename=BUGFIX_FILENAME, content=req.content)
>         await self.repo.docs.save(filename=REQUIREMENT_FILENAME, content="")
>         return AIMessage(
>             content=f"A new issue is received: {BUGFIX_FILENAME}",
>             cause_by=FixBug,
>             send_to="Alex",  # the name of Engineer
>         )
> ```
> 
> `metagpt/roles/engineer.py`中的`_think`:
> 
> ```python
>         if self.config.inc and msg.cause_by in write_plan_and_change_filters:
>             logger.debug(f"TODO WriteCodePlanAndChange:{msg.model_dump_json()}")
>             await self._new_code_plan_and_change_action(cause_by=msg.cause_by)
>             return self.rc.todo
> ```

好的，感谢！

File: metagpt/provider/base_llm.py (hunk 1 above)
Comment (bobooscar) [reply metagpt/provider/base_llm.py:151]: 个人觉得llm的回显信息在问题定位(debug级别）还是很重要的，建议保留。（终端显示默认不打印）

**Pull request closed without merging**